src/schemas/ProgramDonorPublishedAnalysisByDateRange/resolvers/programDonorPublishedAnalysisByDateRange.ts

```diff
@@ -0,0 +1,124 @@
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import esb from 'elastic-builder';
+import { GlobalGqlContext } from 'app';
+import { GraphQLFieldResolver } from 'graphql';
+import {
+  BaseQueryArguments,
+  DonorFields,
+  EsAggs,
+  ProgramDonorGqlResponse,
+} from './types';
+import { Client } from '@elastic/elasticsearch';
+import { ELASTICSEARCH_PROGRAM_DONOR_DASHBOARD_INDEX } from 'config';
+import { UserInputError } from 'apollo-server-express';
+import { convertStringToISODate, validateISODate } from 'utils/dateUtils';
+import { ELASTICSEARCH_DATE_TIME_FORMAT } from '../../../constants/elasticsearch';
+import { differenceInDays, sub as subDate, formatISO } from 'date-fns';
+import logger from 'utils/logger';
+
+const ERROR_TITLE = 'ProgramDonorPublishedAnalysisByDateRange:'
+
+const programDonorPublishedAnalysisByDateRangeResolver: (
+  esClient: Client,
+) => GraphQLFieldResolver<
+  unknown,
+  GlobalGqlContext,
+  BaseQueryArguments & {
+    bucketCount: number;
+    dateRangeFrom: string;
+    dateRangeTo: string;
+    donorFields: DonorFields[];
+  }
+> = esClient => async (source, args, context): Promise<ProgramDonorGqlResponse[]> => {
+  const {
+    bucketCount,
+    dateRangeTo,
+    dateRangeFrom,
+    donorFields,
+    programShortName
+  } = args;
+
+  const areDatesValid = validateISODate(dateRangeFrom) && validateISODate(dateRangeTo);
+  if (!areDatesValid) {
+    throw new UserInputError(`${ERROR_TITLE} Dates must be in ISO format`, {
+      dateRangeFrom,
+      dateRangeTo
+    });
+  }
+
+  const isoDateRangeFrom = convertStringToISODate(dateRangeFrom);
+  const isoDateRangeTo = convertStringToISODate(dateRangeTo);
+
+  const daysInRange = differenceInDays(isoDateRangeTo, isoDateRangeFrom);
+  if (daysInRange < bucketCount) {
+    throw new UserInputError(`${ERROR_TITLE
+      } Days in range must be greater than or equal to bucket count`, {
```

Comment:
What happens if bucket count is negative?

Reply:
I mean, it will error anyways lol... not sure if you want an explicit check for this or not.

Reply:
good idea. I added errors for 

- bucketCount < 1
- daysInRange < 1 (which would mean the FROM date is equal to or later than the TO date)

these both error out anyways but it's good to have specific descriptive errors.